wicketd/src/update_planner.rs

```diff
@@ -0,0 +1,489 @@
+        let update_logs = Mutex::default();
+        let mgs_client = make_mgs_client(log.clone(), mgs_addr);
+
+        // We don't expect to be able to create an update plan before a TUF
+        // repository has been uploaded, and presumably we're being created
+        // before such a thing could happen. Therefore, we store the `Result` of
+        // creating the update plan; if someone tries to call `start()` before
+        // we've been able to successfully generate a plan, we have an error
+        // ready to hand them.
+        let update_plan = StdMutex::new(UpdatePlan::new(&artifact_store));
+
+        Self {
+            mgs_client,
+            artifact_store,
+            update_plan,
+            log,
+            running_updates,
+            update_logs,
+        }
+    }
+
+    pub(crate) fn regenerate_plan(&self) -> Result<(), UpdatePlanError> {
+        match UpdatePlan::new(&self.artifact_store) {
+            Ok(plan) => {
+                *self.update_plan.lock().unwrap() = Ok(plan);
+                Ok(())
+            }
+            Err(err) => {
+                *self.update_plan.lock().unwrap() = Err(err.clone());
+                Err(err)
+            }
+        }
+    }
+
+    pub(crate) async fn start(
+        &self,
+        sp: SpIdentifier,
+    ) -> Result<(), UpdatePlanError> {
+        let plan = self.update_plan.lock().unwrap().clone()?;
+
+        let spawn_update_driver = || async {
+            // Get a reference to the update log for this SP...
+            let update_log = Arc::clone(
+                self.update_logs.lock().await.entry(sp).or_default(),
+            );
+
+            // ...and then reset it to the empty state, removing any events or
+            // state from a previous update attempt.
+            *update_log.lock().unwrap() = UpdateLog::default();
+
+            let update_driver = UpdateDriver {
+                sp,
+                mgs_client: self.mgs_client.clone(),
+                update_log,
+                log: self.log.new(o!("sp" => format!("{sp:?}"))),
+            };
+            tokio::spawn(update_driver.run(plan))
+        };
+
+        let mut running_updates = self.running_updates.lock().await;
+        match running_updates.entry(sp) {
+            // Vacant: this is the first time we've started an update to this
+            // sp.
+            Entry::Vacant(slot) => {
+                slot.insert(spawn_update_driver().await);
+                Ok(())
+            }
+            // Occupied: we've previously started an update to this sp; only
+            // allow this one if that update is no longer running.
+            Entry::Occupied(mut slot) => {
+                if slot.get().is_finished() {
+                    slot.insert(spawn_update_driver().await);
+                    Ok(())
+                } else {
+                    Err(UpdatePlanError::UpdateInProgress(sp))
+                }
+            }
+        }
```

Comment:
In what situations would we want an update to be started here after another one finishes?

Reply:
In practice, I think only if we uploaded a new TUF repository and wanted to MUPdate to an even newer version than we had previously MUPdated, but since wicketd runs "forever" that seems plausible (if not all that likely).

In development, it's handy to be able to start updates even if you've previously updated the target.

Reply:
Makes sense.